In [36]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import os
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split,cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score,mean_absolute_error,mean_squared_error
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')
import math
import random
from tensorflow import keras 
from keras.models import Sequential 
from keras import Input 
from keras.layers import Dense 

In [2]:
data = pd.read_csv('cleaned_used_device_data.csv')
x = data.drop('normalized_used_price',1)
y = data['normalized_used_price']

In [19]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=28)
print(x_train)
print(y_train)
print(x_test)
print(y_test)

      device_brand        os  screen_size        4g        5g  rear_camera_mp  \
1441      4.423648  4.424128        20.32  4.573267  4.372292            5.00   
3072      4.683326  4.424128        16.84  4.573267  5.178265           13.00   
1869      4.459554  4.424128        15.29  4.573267  4.372292           13.00   
3144      4.692627  4.424128        15.80  4.573267  4.372292           12.00   
1508      4.423648  4.424128        25.43  3.949897  4.372292            5.00   
...            ...       ...          ...       ...       ...             ...   
3104      4.459554  4.424128        15.37  4.573267  4.372292           13.00   
2036      4.775672  4.424128        12.88  4.573267  4.372292           13.00   
1302      4.690155  4.424128        12.70  3.949897  4.372292            8.00   
1541      4.423648  4.424128        10.16  3.949897  4.372292            3.15   
1281      4.690155  4.424128        15.24  4.573267  4.372292           13.00   

      front_camera_mp  inte

In [4]:
lr = LinearRegression()
lr.fit(x_train,y_train)
pred = lr.predict(x_test)
print('Score:',r2_score(y_test,pred))
print('MAE:',mean_absolute_error(y_test,pred))
print('MSE:',mean_squared_error(y_test,pred))
print('RMSE:',mean_squared_error(y_test,pred, squared=False))

Score: 0.8213144963799904
MAE: 0.18575229170217075
MSE: 0.05723657883506493
RMSE: 0.23924167453657594


## RMSE calculation based on sklearn issues

### follow this url: https://stackoverflow.com/questions/17197492/is-there-a-library-function-for-root-mean-square-error-rmse-in-python

sklearn >= 0.22.0

sklearn.metrics has a mean_squared_error function with a squared kwarg (defaults to True). Setting squared to False will return the RMSE.

from sklearn.metrics import mean_squared_error

rms = mean_squared_error(y_actual, y_predicted, squared=False)

sklearn < 0.22.0

sklearn.metrics has a mean_squared_error function. The RMSE is just the square root of whatever it returns.

from sklearn.metrics import mean_squared_error
from math import sqrt

rms = sqrt(mean_squared_error(y_actual, y_predicted))



## Design a feed forward neural network
Your designed feed forward neural network performs regression task and compare performance of linear regression and your Neural network performance. Your model can have 2, 3, 4 layers and number of neurons randomly selected from the sample [128, 32, 64, 16]. For example you have 3 layers neural network, you can have 64 neuron in first layer, 32 neurons in second layer, 128 neuron in third layer. 

analyze your results and write the reasons behind getting better or worst performance of your model.

Note: Write a code to randomly select this configuration before creating your own neural network.


In [9]:
layers = [2,3,4]
numLayers = random.randint(0,2)
numLayers = layers[numLayers]
print(numLayers)

2


In [18]:
neurons = [128,32,64,16]
neuronsPicked = True
neuronList = []
count = 0
while neuronsPicked:
    index = random.randint(0,3)
    if (neuronList.count(neurons[index]) == 0):
        neuronList.append(neurons[index])
        count+=1
        if count == numLayers:
            neuronsPicked = False
print(neuronList)

[32, 16]


In [29]:
model = Sequential(name="FFNN")
model.add(Input(shape=(14,), name="Input"))
model.add(Dense(32, kernel_initializer='normal', activation="relu"))
model.add(Dense(16, kernel_initializer='normal', activation="relu"))
model.add(Dense(1))
model.compile(loss="mean_squared_error")

In [33]:
model.fit(
    x_train,
    y_train,
    batch_size=10,
    epochs=32,
    validation_data=(x_test, y_test)
)

Epoch 1/32
228/228 [==============================] - 0s 2ms/step - loss: 0.2047 - val_loss: 0.1327
Epoch 2/32
228/228 [==============================] - 0s 2ms/step - loss: 0.1931 - val_loss: 0.1605
Epoch 3/32
228/228 [==============================] - 0s 2ms/step - loss: 0.1877 - val_loss: 0.1580
Epoch 4/32
228/228 [==============================] - 0s 2ms/step - loss: 0.1741 - val_loss: 0.2765
Epoch 5/32
228/228 [==============================] - 0s 2ms/step - loss: 0.1659 - val_loss: 0.1323
Epoch 6/32
228/228 [==============================] - 0s 2ms/step - loss: 0.1710 - val_loss: 0.1918
Epoch 7/32
228/228 [==============================] - 0s 2ms/step - loss: 0.1626 - val_loss: 0.2135
Epoch 8/32
228/228 [==============================] - 0s 2ms/step - loss: 0.1605 - val_loss: 0.2062
Epoch 9/32
228/228 [==============================] - 0s 2ms/step - loss: 0.1559 - val_loss: 0.1271
Epoch 10/32
228/228 [==============================] - 0s 2ms/step - loss: 0.1498 - val_loss: 0.1414

In [34]:
y_pred = model.predict(x_test)
print('Score:',r2_score(y_test,y_pred))
print('MAE:',mean_absolute_error(y_test,y_pred))
print('MSE:',mean_squared_error(y_test,y_pred))
print('RMSE:',mean_squared_error(y_test,y_pred, squared=False))

Score: 0.4768144675862247
MAE: 0.34021693862567387
MSE: 0.16758690192936906
RMSE: 0.40937379243103617


Linear regression
Score: 0.8213144963799904
MAE: 0.18575229170217075
MSE: 0.05723657883506493
RMSE: 0.23924167453657594

Neural Netowrk
Score: 0.4768144675862247
MAE: 0.34021693862567387
MSE: 0.16758690192936906
RMSE: 0.40937379243103617

The network is likely preforming worse due to limited amounts of training data given the complexity of the data in addition to the network complexity I generated at random likely not having enough complexity. 



Number 4 gradient computation

In [4]:
def computeGrads (a, b , c):
    x = 3 * (a*a) + 2 * (b*b)
    z = (b*b) + c
    dydx = (10 * x)
    print("dy/dx: " + str(dydx))
    dxda = (6 * a)
    print("dx/da: " + str(dxda))
    print("dy/da = dx/da * dy/dx")
    da = dxda * dydx
    print("dy/da: " + str(da))
    dydz = 2
    print("dy/dz: " + str(dydz))
    dzdb = 2 * b
    print("dz/db: " + str(dzdb))
    dydx = 10 * x
    print("dy/dx: " + str(dydx))
    dxdb = 4 * b
    print("dx/db: " + str(dxdb))
    print("dy/db = dx/db * dy/dx + dz/db * dy/dz")
    db = (dxdb * dydx) + (dzdb * dydz)
    print("dy/db: " + str(db))
    dydz = 2
    print("dy/dz: " + str(dydz))
    dzdc = 1
    print("dz/dc: " + str(dzdc))
    print("dy/dc = dz/dc * dy/dz")
    dc = dzdc * dydz
    print("dy/dc: " + str(dc))
    return {da, db, dc}

In [5]:
a = -2
b = 3
c = 10
grads = computeGrads(a, b , c)
print(grads)

dy/dx: 300
dx/da: -12
dy/da = dx/da * dy/dx
dy/da: -3600
dy/dz: 2
dz/db: 6
dy/dx: 300
dx/db: 12
dy/db = dx/db * dy/dx + dz/db * dy/dz
dy/db: 3612
dy/dz: 2
dz/dc: 1
dy/dc = dz/dc * dy/dz
dy/dc: 2
{-3600, 2, 3612}
